### DISCUSSION TITLE
[Important] Fooocus Inpaint V2 Beta Test

### DISCUSSION THREAD

**Discussion by lllyasviel** [lllyasviel/Fooocus#615] [Announcements]
Hello everyone,

Since inpaint is one of the most important features of any image generating software, we plan to step forward and invite you to participant in the **beta test of Fooocus Inpaint V2**.

To test the Fooocus Inpaint V2, you can use `run.bat` to update to at least Fooocus 2.1.33, then go to `advanced -> advanced -> developer debug mode -> Inpaint Engine -> v2`. Then in the first time you use inpaint, it will download a 2.4GB file.

Depending on the results of this beta test (will last one or two weeks, depending on your feedback), some possible outcomes are

* V2 will replace V1 as the default inpaint engine, and the “beta” flag of "Input Image -> Inpaint or Outpaint" will be removed, OR
* V2 will replace V1 as the default inpaint engine, and the “beta” flag remains, OR
* V2 will be removed and we will need to make some modifications, OR
* Nothing will happen.

Thank you for your support and enjoy the new inpaint model.

**Comment by ke1ne** [comment]
V2 engine was downloaded, inpaint worked, but the "beta" flag is still here.

**Reply by lllyasviel** [reply]
"possible outcomes" means one of these things will happen after the beta test (probably will take one or two weeks) depending on user feedbacks

**Reply by ke1ne** [reply]
@lllyasviel oops, considered that as a question. thanks!

**Comment by sunward8888** [comment]
![1](https://github.com/lllyasviel/Fooocus/assets/128116660/2ae02874-152e-4fac-ae45-dc7646f1d7bd)
Interestingly, as the image continues to expand, the original image resolution remains unchanged, which also makes the calculation time longer.

**Reply by sunward8888** [reply]
However, 1 or 2 expansions are enough.

**Comment by lllyasviel** [comment]
it seems that `inpaint v2` still has many problems
probably we will have a `inpaint v2.01` in a few days
`inpaint v1` seems still the most robust method

**Reply by MoonRide303** [reply]
@lllyasviel Might be worth looking at in context of improving inpainting quality: [Designing a Better Asymmetric VQGAN for StableDiffusion](https://arxiv.org/abs/2306.04632):
![image](https://github.com/lllyasviel/Fooocus/assets/130458190/61c46878-9f51-4400-90ac-a4db4ad74b90)

(it was also recently mentioned in [RealFill: Reference-Driven Generation for Authentic Image Completion](https://arxiv.org/abs/2309.16668) from Google).

**Comment by mirek190** [comment]
v1
![2023-10-10_15-49-43_5916](https://github.com/lllyasviel/Fooocus/assets/6817266/16442247-bf40-45ad-9bd7-bc997794abff)
![2023-10-10_15-50-04_8862](https://github.com/lllyasviel/Fooocus/assets/6817266/83023b45-c192-44b6-b993-9107a2cdf89a)

v2

![2023-10-10_15-50-33_2031](https://github.com/lllyasviel/Fooocus/assets/6817266/be9904d2-2cc7-4593-b8d9-79d0c7624239)
![2023-10-10_15-50-53_3753](https://github.com/lllyasviel/Fooocus/assets/6817266/9a56b6f7-9fca-4121-9ddc-8320e6c4f1e9)


v1
![2023-10-10_15-41-15_5383](https://github.com/lllyasviel/Fooocus/assets/6817266/3b9103ae-89b1-4e23-afc6-49b1029e4f0a)

![2023-10-10_15-41-26_8169](https://github.com/lllyasviel/Fooocus/assets/6817266/bf6ed9ca-dddd-4795-b9a4-c29f59bb8043)


v2


![2023-10-10_15-47-09_8791](https://github.com/lllyasviel/Fooocus/assets/6817266/08bf013d-8756-48a3-a2e3-1b4dd6906fc9)
![2023-10-10_15-47-37_6297](https://github.com/lllyasviel/Fooocus/assets/6817266/26402685-cd46-4ae9-b8e0-1057a433ab2e)


v2 seems have more details and more creative but for the time being colors are not match to original picture, some details aloso looks sometimes fuzzy.

**Comment by lllyasviel** [comment]
hello I fixed some problems in inpaint v2 models and now it is called inpaint v2.5
you need at least 2.1.40 to get it

**Comment by lllyasviel** [comment]
it seems that v2.5 is better than v1 in most tests
if users cannot find any cases where v1 is better than v2.5, we will set 2.5 as default in one week

**Comment by mirek190** [comment]
Much better results!   
Seems results from v2.5 are more coherent than for v1 now.

v1

![2023-10-10_23-07-31_5630](https://github.com/lllyasviel/Fooocus/assets/6817266/4dc911e4-2c19-4744-9fa5-cc6f488db431)
![2023-10-10_23-07-44_9303](https://github.com/lllyasviel/Fooocus/assets/6817266/1922dbbe-3660-4669-a44b-77898c727a09)

v2.5

![2023-10-10_23-13-45_4755](https://github.com/lllyasviel/Fooocus/assets/6817266/57b0d306-e4e4-4a48-9dd0-2888286790fa)
![2023-10-10_23-13-56_7912](https://github.com/lllyasviel/Fooocus/assets/6817266/4f5cff77-0498-4328-a7f9-c475b2903857)


v1

![2023-10-10_23-15-05_7655](https://github.com/lllyasviel/Fooocus/assets/6817266/c657c477-dc6d-4578-bc42-9eea8496a22d)
![2023-10-10_23-15-16_3104](https://github.com/lllyasviel/Fooocus/assets/6817266/8f8bf603-cbcd-4542-999a-24ae19733729)


v2.5

![2023-10-10_23-15-43_7990](https://github.com/lllyasviel/Fooocus/assets/6817266/05fcb7e1-7c1c-46ac-9e60-b0b782dbec2f)
![2023-10-10_23-15-54_8582](https://github.com/lllyasviel/Fooocus/assets/6817266/eb381f2c-6aa1-4042-8a72-017c94552305)



Also why when we extending picture the resolution is increasing each time? 
Could you add something as default to resize a final picture because to original one? After fourth time expanding picture it has more than  5000x5000 pixels ...

**Comment by barepixels** [comment]
Ahem, **I like how the resolution increases**.  Think of it as free upscale.  mirek190, you can scale down your image and reload it back into inpaint box.  Don't punish us who want to make prints or wallpapers.  Anyway, v2.5 works great. The newly extended area stays consistently sharp!!!

original 896x1088:
![1-S](https://github.com/lllyasviel/Fooocus/assets/92839033/6d4103fa-6b29-4abe-bdb7-0e20cd42fc9a)

final 1874x1088:
![2-S](https://github.com/lllyasviel/Fooocus/assets/92839033/2204f8a7-72ee-44aa-948a-1a93757f87b9)

version 1, blurry to the left of her hair:
![blurry](https://github.com/lllyasviel/Fooocus/assets/92839033/59df4a89-d3e8-449e-9844-4d437827bf31)


Thank you @lllyasviel

**Reply by mirek190** [reply]
I don't punish you but right now average person and people who want extend picture automatically and more than once quickly find out that process will be totally broken..

Maybe default settings should keep original resolution of the picture after extending and can be extra option under that tab like "extending picture also extend resolution"

**Reply by barepixels** [reply]
How do you know what the average person wants?  
InvokeAI outpainting DOES NOT keep original resolution
Photoshop outpainting DOES NOT keep original resolution
My vote is to keep it like it is.  Less works for lllyasviel

**Reply by mirek190** [reply]
I prefer like is doing midjouarnely.
After outpaintig changing resolution so I can continue outpaintig.
Here I can't because high resolution breaks outpaintig quickly.
So I have to manually changing resolution each time .

**Reply by barepixels** [reply]
We have a solution.  If it starts getting too big, resize the output image to half or 2/3 then reload into inpaint window.    You don't have to do this each time.  Only when it gets too big for your computer.
 
I have a 3070 8gig.  I did 1x to left, batch of 4.  Pick the best then reload. Repeat the process 2x to right.  No problem with speed.  Happy with the size increase.

**Reply by lllyasviel** [reply]
fooocus uses Adobe Firefly’s algorithm to handle this - if you try outpaint many times, you will find that images are bigger but processing speed and iter/s do not decrease. That algorithm proved to be good in their Generative Fill

**Reply by lllyasviel** [reply]
if you find any bigger image causes slower inpaint - please make an issue

**Reply by barepixels** [reply]
So mirek190, you claimed: "Here I can't because high resolution breaks outpaintig quickly." Is this true?

**Reply by mirek190** [reply]
Yes

That tool is amazing without the question ;) but still can be improved ;)

Look on original picture outpaited x3 times (up to 3x times usually works ok  with outpaint v1) ... not counting on my rtx 3090 next outpainting takes 1 minute ... ( 5582x3384 )

outpaint to 9098x5862 ...

inpaint v1 

![2023-10-11_10-02-42_2742](https://github.com/lllyasviel/Fooocus/assets/6817266/232ed6a0-2a92-4480-a800-37a42a5e752b)


then look on the picture I resized before outpainying ( 1395x916) 

outpaint to 2273x1464

 
![25-percent-x2-2023-10-11_10-05-17_7477](https://github.com/lllyasviel/Fooocus/assets/6817266/520a6ade-095c-4bb0-95b7-4426519c90f7)


On the other hand much worse outpaint results I have with version 2.5:
No resizing outpaints to each time is bigger and bigger 
 
first attempt 

![2023-10-10_22-44-29_4928](https://github.com/lllyasviel/Fooocus/assets/6817266/182df748-2b63-4be2-bb47-25b03cb3b29d)
![2023-10-11_09-52-45_2900](https://github.com/lllyasviel/Fooocus/assets/6817266/74141def-14bc-495b-a298-2df8dda84b3d)
![2023-10-11_09-53-11_9696](https://github.com/lllyasviel/Fooocus/assets/6817266/19dd2e04-63d8-43ee-9cfe-6f9091e9bb3e)
![2023-10-11_09-53-48_4459](https://github.com/lllyasviel/Fooocus/assets/6817266/c461e9b5-e567-4cf5-9287-1ca53de2d429)
![2023-10-11_09-55-10_6986](https://github.com/lllyasviel/Fooocus/assets/6817266/ab5c0604-9342-4a75-8e4e-47c3b0afa141)


second attempt
![2023-10-10_22-44-29_4928](https://github.com/lllyasviel/Fooocus/assets/6817266/2e477091-a6d2-4d57-b041-0f8b1d89359c)
![2023-10-11_09-56-04_5160](https://github.com/lllyasviel/Fooocus/assets/6817266/5a16eb26-3217-46d4-ba37-425c1015ce0e)
![2023-10-11_09-56-31_6341](https://github.com/lllyasviel/Fooocus/assets/6817266/9e8b1045-dd24-4d40-837e-a8a6d4fe93c4)
![2023-10-11_09-57-10_3206](https://github.com/lllyasviel/Fooocus/assets/6817266/8cd4b1dc-51d1-48be-94cc-f16823b74c2a)
![2023-10-11_09-58-23_1242](https://github.com/lllyasviel/Fooocus/assets/6817266/37420e5b-ec6a-44b7-84c2-3b4789a81eab)


 version of outpaint v1

first attempt
![2023-10-10_22-44-29_4928](https://github.com/lllyasviel/Fooocus/assets/6817266/54bc3e04-3a91-4846-8cc6-afb59f824259)
![2023-10-11_09-44-09_3677](https://github.com/lllyasviel/Fooocus/assets/6817266/5ae1291d-f5e7-445d-8cee-fdb8cf16f2b4)
![2023-10-11_09-44-38_6523](https://github.com/lllyasviel/Fooocus/assets/6817266/33d50573-a97e-4429-a2f5-4c7b9bf71198)
![2023-10-11_09-45-19_6707](https://github.com/lllyasviel/Fooocus/assets/6817266/29eb9780-f28a-444d-8e1a-a577ca0dcff3)
![2023-10-11_09-46-38_3102](https://github.com/lllyasviel/Fooocus/assets/6817266/4a993ea8-847f-4c41-ab2f-03c10f2734dd)
![2023-10-11_09-49-32_5537](https://github.com/lllyasviel/Fooocus/assets/6817266/c1c3c41d-5f69-4f9a-b2b3-5053e7660fd9)


second attempt
![2023-10-10_22-44-29_4928](https://github.com/lllyasviel/Fooocus/assets/6817266/0fab205c-96ba-421b-93e5-0fd1f5c49719)
![2023-10-11_10-00-24_7012](https://github.com/lllyasviel/Fooocus/assets/6817266/fb7f004f-482d-4821-a6d8-ef44713cae25)
![2023-10-11_10-00-49_3587](https://github.com/lllyasviel/Fooocus/assets/6817266/a9448a16-f66d-409d-bf53-e54d8e7e7a04)
![2023-10-11_10-01-27_7697](https://github.com/lllyasviel/Fooocus/assets/6817266/55e1b89c-012b-468e-8d76-9a3b28b7b92b)
![2023-10-11_10-02-42_2742](https://github.com/lllyasviel/Fooocus/assets/6817266/54ff85a0-59ed-4585-bd45-e9336271cfbb)



My observation 

Seems v2.5 has very high tendency to make frames and later brakes picture completely ( if is a high resolution - resized to original resolution seems less possibility to make frames and breaks outpainting ) 

I hope you understand it better now what I wanted to explain and show. 
That's why I suggest to resize picture to original resolution before outpainting or at least keep resolution below 2K during outpanting ( such behavior should be as default. 
Of course could be extra option under outpainting tab to keep full resolution during that processes. ( as optional not as default)

**Comment by barepixels** [comment]
Am just throwing this here for discussion...  A recent conversation with @MoonRide303 

video link: https://youtu.be/VJymbJ97WDQ?si=TYNKhARgmvKvOEi2
![Left_Right_Top_Bottom) ](https://github.com/lllyasviel/Fooocus/assets/92839033/3cbb34d1-75e9-48be-b092-5cb727f9b805)


Ideally, the total new area for outpainting should be around 1024x1024.  

I don't have the math formula, but if you are going to do (top, bottom, left, right) at the same time, maybe to optimize results, calculate how much you have to downsize so the newly generated area is around 1024x1024 pixel count
![new square](https://github.com/lllyasviel/Fooocus/assets/92839033/6f9d8fa4-0f8c-478e-975f-719dcc5940dc)

Maybe that is why the below outpainting came out good.  
1st, Left new area = approx 326x1088
2nd, Right new area = approx 326x1088
3rd, Right new area = approx 326x1088
Perhaps I should upscale the original a little so each of the 3 tiles has a similar pixel count as in a (1024x1024) area.  Resulting a larger final image without sacrificing quality
![2-S](https://github.com/lllyasviel/Fooocus/assets/92839033/936eaf76-0073-41b8-b1c8-c49e9fa1a493)

It is not about the total size of your image.  The fooocus ;) should be on the size of the tile.

**Reply by mirek190** [reply]
Thank you! 
So I was right from the beginning just from by my own tests ;)
Image have to be resized before outpaint bigger area than 1024x1024.
Or maybe outpaint could work in the tile method ( tile by tile  ) ....

**Reply by barepixels** [reply]
and I was right. It is not about the total size of your image ;)
Absolutely DO NOT default to resize a final picture to original one

**Reply by mirek190** [reply]
It could generate tile by tile with outpaiting ... but that takes a lot time.
 I think increasing the picture size we should leave under upscaler not under outpaintig if picture is bigger than 1024x1024

**Reply by barepixels** [reply]
The current upscaler is not that great.  No tile

**Comment by barepixels** [comment]
v2.5 Inpaint clouds
![3-S](https://github.com/lllyasviel/Fooocus/assets/92839033/a2ce0983-b78c-4d05-9652-5db7bdf770bb)

**Comment by batfruit** [comment]
Testing 2.5 inpaint

Struggling with image out-paint especially feet, please see image.   Often by the time you reach the feet portion they are never flat on the floor and mutated / bad anatomy.  Often the lower limbs get a bit extended disproportionate, like x2 calf's to ankle and elongated foot.

Also thin horizontal seam fringe as per screenshot.

Thanks for all your superb work , fooocus is my favorite!

![image](https://github.com/lllyasviel/Fooocus/assets/127588424/37a9026c-c787-4c23-b71a-10dcb9a26f3e)

**Reply by barepixels** [reply]
see this https://github.com/lllyasviel/Fooocus/issues/647

**Comment by 8Docent8** [comment]
Hello. I really like Fooocus and the new version of "Inpaint", which is just great for helping photographers. I didn't know where to ask, so I'll ask here. Are you planning to combine "Inpaint" and "Image Prompt" in the future? Is it even possible to combine these two tools? Anyway, thanks for the work done.

**Reply by lllyasviel** [reply]
It is hidden deep as not production-ready

Advanced->Advanced->Debug mode->Developer Tools->Debug Control->Mixing Inpaint and Image Prompt

**Comment by batfruit** [comment]
awesome thank you!
________________________________
